### ISSUE TITLE
read_json not being determined as a Map

### ISSUE THREAD

**Issue by illusional** [chanzuckerberg/miniwdl#320]
I'm using read_json to read the contents of a file with a json output, and then immediately trying to access a property on this (through the following minimal example):

```
version development
task readjson {

  command <<<
      echo '{"out": "Hello, WDL!"}' > out.txt
  >>>

  output {
    String out = read_json('out.txt')["out"]
  }
}
```

This validates and runs correctly in Cromwell, but in miniWDL I get the following error:

```
miniwdl run --verbose codetool.wdl
(codetool.wdl Ln 10 Col 18) Not an array
        String out = read_json('out.txt')["out"]
                     ^^^^^^^^^^^^^^^^^^^^
```

I've noticed that you can coerce it to the correct type which will then validate, eg:
```
version development
task readjson {

  command <<<
      echo '{"out": "Hello, WDL!"}' > out.txt
  >>>

  output {
    Map[String, String] coercedmap = read_json('out.txt')
    String out = coercedmap["out"]
  }
}
```

But this has the disadvantage that it returns more than what's required.

---

Versions:
- `v0.6.3` (through Pip)
- Latest on Git: `miniwdl v0.6.4.dev14`

**Comment by illusional** [comment]
Hey @mlin, I've got a similar error. I'm happy to open up a new error for this but thought this might be linked. This time, instead I'm outputting an `Array[String]` instead of just a `String`, I get the following error:

```
2020-03-27 11:22:10,022 wdl.t:readjson ERROR task readjson (readtest.wdl Ln 3 Col 1) failed :: dir: "<outputdir>", error: "EvalError", message: "function evaluation failed, couldn't construct Any from input [\"Element 1\", \"Element 2\"]", node: "decl-out", pos: {"source": "/path/to/readtest.wdl", "line": 10, "column": 18}
2020-03-27 11:22:10,023 miniwdl-run ERROR task readjson (readtest.wdl Ln 3 Col 1) failed :: error: "EvalError", message: "function evaluation failed, couldn't construct Any from input [\"Element 1\", \"Element 2\"]", node: "decl-out", pos: {"source": "/path/to/readtest.wdl", "line": 10, "column": 18}, dir: "<outputdir>"
```

**readtest.wdl**

```wdl
version development

task readjson {

  command <<<
      echo '{"out": ["Element 1", "Element 2"]}' > out.txt
  >>>

  output {
    String out = read_json('out.txt')["out"]
  }
}
```

**Comment by mlin** [comment]
Thanks -- yea, looks like [I punted on this case](https://github.com/chanzuckerberg/miniwdl/pull/330/files#diff-86d4b2231a3995219f574f6892b388efR331) previously. Will take another look

**Comment by illusional** [comment]
Thanks for the quick reply! No stress at all :)

**Comment by illusional** [comment]
Hey @mlin, I'm not familiar with the repo so much but I looked at where it tries to inspect the type and just added an extra layer of inspection. For arrays, it is expecting homogeneity. 

- https://github.com/illusional/miniwdl/pull/2 (addressed against my own master to avoid a PR showing up here)

String wasn't in that block, so I added it because it made my thing work and it looked like it belong.

**Comment by mlin** [comment]
@illusional the solution in just-released v0.7.1 is very much along the lines you suggested, but exapts some type unification logic that had existed elsewhere for other purposes. There will probably still be some weird corner cases like non-homogeneity in lists and object value types, as you mention -- let us know if you hit anything that clearly ought to work even given that JSON isn't 100% embeddable in WDL values. Thanks!
